### PR TITLE
WiP http: add lazy map for eliminating O(n) header operations.

### PR DESCRIPTION
**Pull request by jmarantz**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by jmarantz** [comment]
Speed-test comparison with no maps, with a flat_hash_map, and with a multi_map. To me it looks like with a reasonable # of headers, multi_map is the best choice.

Screen Shot 2019-12-07 at 9 23 39 AM

**Comment by jmarantz** [comment]
An easier to read graph covering just the benchmarks that have O(N) growth with no map:

Screen Shot 2019-12-07 at 1 32 00 PM

**Comment by mattklein123** [comment]
@jmarantz do you mind summarizing what we are doing/testing here? Also, as an aside, did you try the absl btree maps? They should be faster than the std maps AFAIK.

**Comment by jmarantz** [comment]
I don't mind :) This is not really in shape to be a mergeable PR as it makes things strictly more complex. But I wanted to think quantitatively about the overhead have maintaining maps. FWIW we do see HeaderMap stuff show up in flame-graphs and it's not the lowest hanging fruit right now but it does come up. I suspect in our environment we have a bunch of non-O(1) headers, especially Set-Cookie which needs to stay O(N) as that's currently tied to coalescing.

I haven't tried -- and didn't know about -- absl::btree . But one more thing is that the multi-map variant in this draft-PR could also allow for a fast (O(log n) + O(# matches)) impl of removePrefix(), though I don't know specifically if that's coming up as an issue.

I thing my thoughts on plausible end-states include:
 * drop this PR; e.g. assume there's never enough non-O(1) headers in the real world for this to matter. I suspect 20+ response headers would be pretty common but the perf-tests don't yet cover that point.
 * accept the complexity of one of these options (we should try btree too) in exchange for removing the O(N) exposure, with much fuzzing to give us confidence.
 * solving #4815 by updating the trie on startup and dropping this PR; basically allowing any extension to easily add O(1) headers, per discussion on that issue.

**Comment by mattklein123** [comment]
OK thanks makes sense to me. I think this is a worthwhile thing to investigate for sure.

File: source/common/http/header_map_impl.h [hunk 1]
```diff
@@ -99,6 +110,16 @@ class HeaderMapImpl : public HeaderMap, NonCopyable {
   void dumpState(std::ostream& os, int indent_level = 0) const override;
 
 protected:
+  struct HeaderEntryImpl;
+  using HeaderNode = std::list<HeaderEntryImpl>::iterator;
+#if HEADER_MAP_USE_FLAT_HASH_MAP
+  using HeaderNodeVector = std::vector<HeaderNode>;
```
Comment (mattklein123) [inline source/common/http/header_map_impl.h]: I would try https://github.com/abseil/abseil-cpp/blob/master/absl/container/inlined_vector.h here with a default size of 1 which would be the common case?

File: source/common/http/header_map_impl.cc [hunk 2]
```diff
@@ -578,7 +602,42 @@ HeaderMap::Lookup HeaderMapImpl::lookup(const LowerCaseString& key,
...
+      lazy_map_.insert(std::make_pair(key, node));
+#endif
+    }
+  }
+  return true;
+}
+
+HeaderMapImpl::HeaderLazyMap::iterator
+HeaderMapImpl::HeaderList::find(absl::string_view key) const {
+  if (lazy_map_.empty()) {
```
Comment (mattklein123) [inline source/common/http/header_map_impl.cc]: Should this also use maybeMakeMap() to be consistent and also share the construction code?

**Comment by mattklein123** [comment]
Cool stuff just left some random comments from a quick read, will look more later. :)

File: source/common/http/header_map_impl.h (hunk 1 above)
Comment (jmarantz) [reply source/common/http/header_map_impl.h]: good idea; I'll give that a try.

File: source/common/http/header_map_impl.cc (hunk 2 above)
Comment (jmarantz) [reply source/common/http/header_map_impl.cc]: Yeah this could use a refactor; I'll clean that up.

**Comment by mattklein123** [comment]
@jmarantz what are your thoughts on next steps for this PR?

/wait-any

**Comment by jmarantz** [comment]
Will be moving forward with this when I get some coding time :)

I started experimenting with your suggestions and the inlined-vector improved things a bit, the btree didn't seem to help, but I would not say my experiments are conclusive.

**Comment by mattklein123** [comment]
OK I will put this into waiting for now with no stalebot. Let us know when you feel ready to have a conversation about whether we want to go with this approach or not.

**Pull request closed without merging**